bukkit/src/main/java/com/viaversion/viaversion/bukkit/listeners/JoinListener.java

```diff
@@ -0,0 +1,120 @@
+import com.viaversion.viaversion.bukkit.handlers.BukkitEncodeHandler;
+import com.viaversion.viaversion.bukkit.util.NMSUtil;
+import io.netty.channel.Channel;
+import org.bukkit.entity.Player;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.EventPriority;
+import org.bukkit.event.Listener;
+import org.bukkit.event.player.PlayerJoinEvent;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.logging.Level;
+
+public class JoinListener implements Listener {
+
+    private static final Method getHandle;
+    private static final Field connection;
+    private static final Field networkManager;
+    private static final Field channel;
+    private static final boolean enabled;
+
+    static {
+        Method gh;
+        Field conn, nm, ch;
+        boolean en = true;
+        try {
+            gh = NMSUtil.obc("entity.CraftPlayer").getDeclaredMethod("getHandle");
+            conn = findField(gh.getReturnType(), "PlayerConnection");
+            nm = findField(conn.getType(), "NetworkManager");
+            ch = findField(nm.getType(), "Channel");
+        } catch (NoSuchMethodException | NoSuchFieldException | ClassNotFoundException e) {
+            en = false;
+            gh = null;
+            conn = nm = ch = null;
+
+            Via.getPlatform().getLogger().log(
+                    Level.WARNING,
+                    "Couldn't find reflection methods/fields to access Channel from player.\n" +
+                            "Login race condition fixer will be disabled.\n" +
+                            " Some plugins that use ViaAPI on join event may work incorrectly.", e);
+        }
+
+        getHandle = gh;
+        connection = conn;
+        networkManager = nm;
+        channel = ch;
+        enabled = en;
+    }
+
+    // Loosely search a field with any name, as long as it matches the type
+    private static Field findField(Class<?> cl, String type) throws NoSuchFieldException {
+        for (Field field : cl.getDeclaredFields()) {
+            if (field.getType().getSimpleName().equals(type))
+                return field;
+        }
+        throw new NoSuchFieldException(type);
+    }
+
+    @EventHandler(priority = EventPriority.LOWEST)
+    public void onJoin(PlayerJoinEvent e) {
+        if (!enabled) return;
+        Player player = e.getPlayer();
+
+        UserConnection user = getUserConnection(player);
+        if (user == null) {
+            Via.getPlatform().getLogger().log(Level.WARNING,
+                    "Could not find UserConnection for logging-in player {0}",
+                    player.getUniqueId());
+            return;
+        }
+
+        ProtocolInfo info = user.getProtocolInfo();
+        info.setUuid(player.getUniqueId());
+        info.setUsername(player.getName());
+        Via.getManager().getConnectionManager().onLoginSuccess(user);
+    }
+
+    private UserConnection getUserConnection(Player player) {
+        Channel channel = getChannel(player);
```

Comment:
Check if the channel is still open

Reply:
what thread does the channel close in, wouldn't that be grounds for a new possible race condition?

User logs in, and in network thread, connection closes. Later the PlayerJoinEvent (on main thread) re-sets the same player.
If channel closing is in fact in network thread, checking here for open only reduces how likely it is that we find that race condition, but we may still get such a case

Reply:
It can be closed from anywhere, but we have a close future listener that includes already closed channels, so that doesn't matter

Reply:
yeah, but nothing prevents us re-adding (by calling onLoginSuccess) *after* it had already closed, leaking the UserConnection, this could happen if main thread is laggy enough that on network thread you fully joined and disconnected, and you hit the timings *just right* in a very obscure way

Reply:
No, the close future is executed when set and the channel is already closes, immediately removing it again. but that'd be alleviated by the latest review in any case